src/func.hpp

```diff
@@ -0,0 +1,174 @@
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "environment.hpp"
+#include "type.hpp"
+#include "wassert.hpp"
+
+namespace winter {
+
+class ModuleInstance;
+class ModuleInstanceInternal;
+
+struct UnlinkedFuncInternal;
+class UnlinkedFunc;
+
+struct LinkedFuncInternal;
+class LinkedFunc;
+
+using JitFunction = uint32_t (*)(LinkedFuncInternal* func);
+
+class InstructionStream {
+    std::vector<uint8_t> _stream;
+public:
+    InstructionStream(std::vector<uint8_t>&& stream) : _stream(stream) {}
+
+    size_t size() const { return _stream.size(); }
+
+    friend class InstructionCursor;
+};
+
+class InstructionCursor {
+    const InstructionStream* _stream;
+    const uint8_t* _cursor;
+
+    const uint8_t* begin() const { return _stream->_stream.data(); }
+    const uint8_t* end() const { return _stream->_stream.data() + _stream->size(); }
+public:
+    InstructionCursor(const InstructionStream* stream, size_t off) : _stream(stream) {
+        WASSERT(off <= _stream->size(), "Instruction cursor out-of-bounds");
+        _cursor = _stream->_stream.data() + off;
+    }
+
+    size_t off() const { return _cursor - _stream->_stream.data(); }
+
+    void jump_relative(ssize_t off) {
+        WASSERT(off >= begin() - _cursor, "Instruction cursor out-of-bounds");
+        WASSERT(off <= end() - _cursor, "Instruction cursor out-of-bounds");
+
+        _cursor += off;
+    }
+
+    uint8_t read_u8() {
+        WASSERT(_cursor != end(), "Instruction cursor out-of-bounds");
+
+        return *(_cursor++);
+    }
+};
+
+struct AbstractFunc {
+    bool is_import;
+
+    std::string debug_name;
+    std::shared_ptr<const InstructionStream> instrs;
+    FuncSig sig;
+
+    AbstractFunc(bool is_import, std::string debug_name, std::shared_ptr<const InstructionStream> instrs, FuncSig sig)
+        : is_import(is_import), debug_name(std::move(debug_name)), instrs(std::move(instrs)), sig(std::move(sig)) {}
+
+    static AbstractFunc for_import(FuncSig sig) {
+        return AbstractFunc(true, "", nullptr, std::move(sig));
+    }
+};
+
+struct UnlinkedFuncInternal {
+    JitFunction jit_fn = nullptr;
+    const FuncSig* sig;
+
+    UnlinkedFunc* container;
+};
+
+class UnlinkedFunc {
+    UnlinkedFuncInternal _internal;
+    std::string _debug_name;
+    std::shared_ptr<const InstructionStream> _instrs;
+
+    UnlinkedFunc() {
+        _internal.container = this;
+    }
+    UnlinkedFunc(const UnlinkedFunc&) = delete;
+
+    UnlinkedFunc& operator=(const UnlinkedFunc&) = delete;
+public:
+    UnlinkedFuncInternal* internal() { return &_internal; }
+
+    const FuncSig& signature() const { return *_internal.sig; }
+    const std::string& debug_name() const { return _debug_name; }
+    const std::shared_ptr<const InstructionStream> instrs() const { return _instrs; }
+
+    static std::shared_ptr<UnlinkedFunc> instantiate(const AbstractFunc& func, Environment& env);
+    static std::shared_ptr<UnlinkedFunc> mock(const FuncSig* sig) {
```

Comment:
This (and the other mysterious `mock` function) are for creating fake functions to be used in the module linking tests... A function created from here has just enough information filled in to be able to link, but not enough to be able to actually run. I'd love to hide these away, but they need to be available to the test code so they can't be made private, and since the test code is in an anonymous namespace, I can't use a friend declaration to give it access.

That being said, I really don't like the name of this function, but I have yet to come up with a better name for it...

Reply:
Maybe `create_mock()`?

Reply:
Renamed.